backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/Boxing.kt

```diff
@@ -50,3 +53,166 @@ internal fun KonanSymbols.getTypeConversion(
 internal fun KonanSymbols.getUnboxFunction(valueType: ValueType): IrSimpleFunctionSymbol =
         this.unboxFunctions[valueType]
                 ?: this.boxClasses[valueType]!!.getPropertyGetter("value")!! as IrSimpleFunctionSymbol
+
+
+/**
+ * Represents static array of boxes.
+ */
+
+internal enum class BoxCache(val valueType: ValueType) {
+    BYTE(ValueType.BYTE),
+    SHORT(ValueType.SHORT),
+    CHAR(ValueType.CHAR),
+    INT(ValueType.INT),
+    LONG(ValueType.LONG);
+
+    private val valueTypeName = valueType.name.toLowerCase().capitalize()
+    private val getIntrinsic = "konan.internal.getCached${valueTypeName}Box"
+    private val checkIntrinsic = "konan.internal.in${valueTypeName}BoxCache"
+
+    val cacheName = "${valueTypeName}Boxes"
+    val rangeStartName = "${valueTypeName}RangeStart"
+    val rangeEndName = "${valueTypeName}RangeEnd"
+
+    companion object {
+        /**
+         * returns cache corresponding to the given getIntrinsic name.
+         */
+        fun getCacheByBoxGetter(getBoxMethodName: String): BoxCache? =
+            BoxCache.values().firstOrNull { it.getIntrinsic == getBoxMethodName }
+
+        /**
+         * returns cache corresponding to the given checkIntrinsic name.
+         */
+        fun getCacheByInRangeChecker(checkRangeMethodName: String): BoxCache? =
+            BoxCache.values().firstOrNull { it.checkIntrinsic == checkRangeMethodName }
+
+        /**
+         * Initialize globals.
+         */
+        fun initialize(context: Context) {
+            values().forEach {
+                it.initRange(context)
+                it.initCache(context)
+            }
+        }
+    }
+}
+
+/**
+ * Adds global that refers to the cache.
+ * If output target is native binary then the cache is created.
+ */
+private fun BoxCache.initCache(context: Context): LLVMValueRef {
+    return if (context.config.produce.isNativeBinary) {
+        context.llvm.staticData.createBoxes(this)
+    } else {
+        context.llvm.staticData.addGlobal(cacheName, getLlvmType(context), false)
+    }
+}
+
+/**
+ * Creates globals that defines the smallest and the biggest cached values.
+ */
+private fun BoxCache.initRange(context: Context) {
+    if (context.config.produce.isNativeBinary) {
+        val (start, end) = getRange(context)
+        // Constancy of these globals allows LLVM's constant propagation and DCE
+        // to remove fast path of boxing function in case of empty range.
+        context.llvm.staticData.placeGlobal(rangeStartName, createConstant(start), true)
+                .setConstant(true)
+        context.llvm.staticData.placeGlobal(rangeEndName, createConstant(end), true)
+                .setConstant(true)
+    } else {
+        context.llvm.staticData.addGlobal(rangeStartName, valueType.llvmType, false)
```

Comment:
Not sure if this is a good approach. Why not just make all that in C++ code, and make LLVM just call C function?